### PR TITLE
Add Mac support for library loading

### PR DESCRIPTION
**Pull request by jasonleenaylor**
Addresses issue #141

### CODE REVIEW

File: source/icu.net/NativeMethods/NativeMethods.cs [hunk 1]
```diff
@@ -81,7 +81,26 @@ private static string dlerror()
 			var ptr = _dlerror();
 			return Marshal.PtrToStringAnsi(ptr);
 		}
+		private const string MAC_LIBDL_NAME = "libdl.dylib";
 
+		[DllImport(MAC_LIBDL_NAME, SetLastError = true)]
+		private static extern IntPtr mac_dlopen(string file, int mode);
```
Comment (ermshiperete) [inline source/icu.net/NativeMethods/NativeMethods.cs:87]: So Mac actually method names prefixed with `mac_` ? I wouldn't have expected that.

**Review by ermshiperete — COMMENTED** [review]
Please add a comment to `CHANGELOG.md` - seems worth mentioning that Mac is now supported, and add `+semver:minor` to the commit message.

If the method names on Mac are the same as on Linux, we should solve this differently:

- change `LIBDL_NAME`: `private const string LIBDL_NAME = "libdl.so";`
- modify `App.config` and add:

  ```
  <configuration>
    <dllmap os="!windows,osx" dll="libdl.so" target="libdl.so.2" />
    <dllmap os="osx" dll="libdl.so" target="libdl.dylib"/>
  </configuration>
   ```

**Review by ermshiperete — APPROVED** [review]
:lgtm:

**Comment by chrisvire** [comment]
@jasonleenaylor You can remove me from the review. I have never done this so I don't know.

**Comment by imnasnainaec** [comment]
I'm still encountering `System.AggregateException : One or more errors occurred. (Unable to load shared library 'libdl.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(liblibdl.so, 1): image not found)`: https://github.com/sillsdev/TheCombine/issues/927#issuecomment-791757558

**Comment by ermshiperete** [comment]
The problem is that the nuget package is missing the `icu.net.dll.config` file.

**Comment by martin-shields-sage** [comment]
> The problem is that the nuget package is missing the `icu.net.dll.config` file.

I have the same issue as well, even adding the config file, for some reason the DLL map is not working on mac

**Comment by ermshiperete** [comment]
> > The problem is that the nuget package is missing the `icu.net.dll.config` file.
> 
> I have the same issue as well, even adding the config file, for some reason the DLL map is not working on mac

@martin-shields-sage What mono version are you using? What icu-dotnet version? And can you show the exception you're getting including a stack trace?

**Comment by martin-shields-sage** [comment]
Exception Message:
`Unable to load shared library 'libdl.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(liblibdl.so, 1): image not found`

Stack Trace:
`   at Icu.NativeMethods.dlopen(String file, Int32 mode)
   at Icu.NativeMethods.GetIcuLibHandle(String basename, Int32 icuVersion)
   at Icu.NativeMethods.LoadIcuLibrary(String libraryName)
   at Icu.NativeMethods.get_IcuCommonLibHandle()
   at Icu.NativeMethods.u_init(ErrorCode& errorCode)
   at Icu.Wrapper.Init()
   at testLocal.Program.Main(String[] args) in /Users/martin.shields/Projects/testLocal/testLocal/Program.cs:line 14`

Mono Version
Mono JIT compiler version 6.12.0.162 (2020-02/2ca650f1f62 Tue Nov 30 10:18:09 EST 2021)

For some reason it appears to be ignoring the config file which was added and insisting on using libdl.so instead of libdl.dylib. 

Not sure if I am doing something wrong but I ran dotnet pack on the project and used that nuget package which did include the config file

**Comment by ermshiperete** [comment]
Unfortunately I don't have a Mac available, so can't test this.

Make sure you have the `icu.net.dll.config` next to the `icu.net.dll` binary in the same directory as your test app.

**Pull request merged**